Omit restricted resource types in GCP and AWS

MikaelSmith opened the pull request:
GCP and AWS access controls can limit what types of APIs a user has
access to. A common one is to limit users to certain subsets of the API,
such as only Compute or Storage. When that happens, any operations on
the representative directory will fail.

Create a new Inaccessible attribute that entries set when they are
inaccessible due to lack of permissions. Inaccessible entries will be
hidden from list and find commands, with the failure reason logged to
the relevant activity journal.

In practice, you would see
```
$ cd gcp/Storage-Only-Project
$ ls
storage
$ stree
.
├── compute
│   └── ...
└── storage
└── ...
$ whistory
...
7   2019-10-14 16:13  ls
$ whistory 7
Oct 14 16:13:12.752 FUSE: List /gcp/Storage-Only-Project
Oct 14 16:13:13.654 Omitting Storage-Only-Project/compute: googleapi:
Error 403: user@example.com does not have compute.list access to p...
```

Fixes #534.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
@adrienthebo mind kicking the tires on this?

MikaelSmith commented:
I think there are other types it should probably cover, but this seems like one of the major issues.

ekinanp commented:
I think the more general issue here is "Does the user have access to this entry?". That could be captured by a `Restricted`/`Inaccessible` property on the entry, which generalizes well to external plugins because we can make that an (optional) key in the entry JSON (`plugin.CachedList` would have additional code to omit restricted entries). I think we should go with that for this PR, because I don't see `SelectListable` generalizing to external plugins.

Long-term, the more general issue is "What are 'Wash' permissions?" Entry access is one such permission while something like "what are the allowable supported actions" could be another (with each action having its own set of permissions; action-specific permissions could be useful for the upcoming Signal action b/c then we can enforce API-specific "signal" permissions [e.g. you can start/stop a VM but can't hibernate it]).

adrienthebo commented:
LGTM for the specific fix:

```
wash gcp ❯ cd s2-backups                                                                                                                                                                                                                                                    ✔  17:03:23 
wash gcp/s2-backups ❯ ls
dr-xr-x--- adrien adrien 4 KB Tue Oct 15 17:03:15 2019   storage                                                                                                                                                                                                           ✔  17:03:24 
wash gcp/s2-backups ❯
```

@ekinanp raises a good point, though one that might be worth discussing in detail in the larger picture - could one use file permissions to indicate access?

But for this specific issue, this solved the issue that I discovered and I'm happy with the fix.

MikaelSmith commented:
Working on what changing to an Inaccessible attribute might look like.

MikaelSmith committed:
Omit restricted resource types in GCP and AWS

GCP and AWS access controls can limit what types of APIs a user has
access to. A common one is to limit users to certain subsets of the API,
such as only Compute or Storage. When that happens, any operations on
the representative directory will fail.

Create a new inaccessible attribute that entries set when they are
inaccessible due to lack of permissions. Inaccessible entries will be
hidden from list and find commands, with the failure reason logged to
the relevant activity journal.

In practice, you would see
```
$ cd gcp/Storage-Only-Project
$ ls
storage
$ stree
.
├── compute
│   └── ...
└── storage
└── ...
$ whistory
...
7   2019-10-14 16:13  ls
$ whistory 7
Oct 14 16:13:12.752 FUSE: List /gcp/Storage-Only-Project
Oct 14 16:13:13.654 Omitting Storage-Only-Project/compute: googleapi:
Error 403: user@example.com does not have compute.list access to p...
```

Fixes #534.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
I've reworked the implementation. Entries now define a new function to call for a permission check. This moves most of the actual checking internally, so we can enforce some standard interactions.

MikaelSmith commented on plugin/types.go:
```diff
@@ -193,3 +193,11 @@ type Readable interface {
 	Open(context.Context) (SizedReader, error)
 }
+
+// Permissible is used to define how to check that adequate permissions to access the resource are
```
To make this work with external plugins we'd need a Permitted method to check. Obviously that could be statically declared when listing the items.

MikaelSmith replied:
An attribute will be simpler if we're getting rid of the need to test whether an entry is Permissible (or making it an explicit thing a plugin needs to set).

ekinanp commented on cmd/docs.go:
```diff
@@ -37,5 +37,13 @@ func docsMain(cmd *cobra.Command, args []string) exitCode {
+	if schema.PermissibleChildren {
+		// Intentionally adds an empty newline after the provided description.
+		cmdutil.Println(`
+If a subdirectory is listed in 'stree' but not visible in your directory then you are likely
```
I think we should move this over to `stree` and just add a note somewhere in its output. Or move it to the website.

MikaelSmith replied:
So rather than track this, we'll look at including a note about it in documentation on the `stree` command and the website.

MikaelSmith commented on plugin/cache.go:
```diff
@@ -189,7 +190,21 @@ func cachedList(ctx context.Context, p Parent) (map[string]Entry, error) {
 	}
 
-	return cachedEntries.(map[string]Entry), nil
+	// Return a copy of the map so we don't mutate the cached version. Future calls to list should
```
If we're going to use an attribute on the entry to determine whether we have permission to operate on it, deciding whether or not to include that entry in a list of entries can be moved closer to where it's displayed to the user.

MikaelSmith commented:
Updated with the latest take on this using an attribute.

ekinanp commented on plugin/entryAttributes.go:
```diff
@@ -179,6 +180,15 @@ func (a *EntryAttributes) SetSize(size uint64) *EntryAttributes {
 }
 
+// Inaccessible returns true if accessing an entry is expected to fail due to missing permissions.
+func (a *EntryAttributes) Inaccessible() bool {
```
I think this should move over to the `EntryBase` class. Main reason is b/c the `inaccessible` attribute isn't something that you're going to filter on.

MikaelSmith replied:
Making it an attribute made it straight-forward to set from an external plugin.

MikaelSmith replied:
I'll look into how the external plugin should work.

MikaelSmith replied:
Added external plugin support.

MikaelSmith committed:
Implement external plugin support for inaccessible entries

Adds an `inaccessible_reason` attribute on an entry that provides a
mechanism to flag that the entry should not be listed and a reason for
its omission that Wash will log.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented on plugin/entryBase.go:
```diff
@@ -133,6 +135,13 @@ func (e *EntryBase) SetAttributes(attr EntryAttributes) *EntryBase {
 
+// MarkInaccessible sets the inaccessible attribute and logs a message about why the entry is
+// inaccessible.
+func (e *EntryBase) MarkInaccessible(ctx context.Context, err error) {
```
Done.
